opentelemetry/src/sdk/trace/span.rs

```diff
@@ -173,34 +173,49 @@ impl crate::trace::Span for Span {
 impl Span {
     fn ensure_ended_and_exported(&mut self, timestamp: Option<SystemTime>) {
         if let Some(mut data) = self.data.take() {
-            // Ensure end time is set via explicit end or implicitly on drop
-            if let Some(timestamp) = timestamp {
-                data.end_time = timestamp;
-            } else if data.end_time == data.start_time {
-                data.end_time = crate::time::now();
-            }
-
-            // Notify each span processor that the span has ended
             if let Some(provider) = self.tracer.provider() {
-                let mut processors = provider.span_processors().iter().peekable();
-                let resource = provider.config().resource.clone();
-                let mut span_data = Some((data, resource));
-                while let Some(processor) = processors.next() {
-                    let span_data = if processors.peek().is_none() {
-                        // last loop or single processor/exporter, move data
-                        span_data.take()
-                    } else {
-                        // clone so each exporter gets owned data
-                        span_data.clone()
-                    };
-
-                    if let Some((span_data, resource)) = span_data {
-                        processor.on_end(build_export_data(
-                            span_data,
-                            self.span_context.clone(),
-                            resource,
-                            &self.tracer,
-                        ));
+                let processors = provider.span_processors();
+                if processors.is_empty() {
+                    return;
+                }
+
+                // Ensure end time is set via explicit end or implicitly on drop
+                if let Some(timestamp) = timestamp {
+                    data.end_time = timestamp;
+                } else if data.end_time == data.start_time {
+                    data.end_time = crate::time::now();
+                }
+
+                if processors.len() == 1 {
```

Comment:
This might be even more efficient as a match?

```
match processors {
    [] => return,
    [processor] => // export on single processor,
    _ => // send to all processors
}
```

Reply:
good idea 👍